proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2142 @@
+  operators, we introduce a familly of (infix) type constructors:
+  ``(:p->)`` for each multiplicity ``p``. This technically steals
+  syntax as ``(:)`` is a valid type operator under the discouraged
+  ``-XDataKinds`` syntax. But this should not be a problem in
+  practice.
+
+  - In ``(:p->)``, ``p`` can be any type expression of kind
+    ``Multiplicity`` (see below). So that the following is legal
+    (though see Alternatives_):
+
+    ::
+
+      type familly F (a :: *) :: Multiplicity
+      f ::  forall (a :: *). Int  :(F a)-> a -> a
+- Binders can also be annotated with a multiplicity:
+
+  ::
+
+    \x ::('One) A -> x
+
+  is the identity function at type ``A ->. A``. This adds a new syntactic for binders:
+
+  ::
+
+    VAR ::( TYPE_EXPRESSION ) TYPE_EXPRESSION
+
+  where the first ``TYPE_EXPRESSION`` must be of kind ``Multiplicity``
+  (see below). This creates a new lexical token ``::(``, which does
+  not seem to be a problem.
+
+  This form is disallowed for:
+
+  - Type variables
+
+    ::
+
+      forall (a ::('One)). a -> Int -- rejected
+  - Top-level signatures (though, see `Toplevel binders`_)
+
+    ::
+
+      foo ::('One) A -> B -- rejected
+      foo x = …
+
+  The form is however permitted in records (see `Records`_ below)
+
+  ::
+
+    data R = R { unrestrictedField ::(Omega) A, linearField ::(One) B }
+
+In the fashion of levity polymorphism, the proposal introduces a data
+type ``Multiplicity`` which is treated specially by the type checker,
+to represent the multiplicities:
+
+- ::
+
+    data Multiplicity
+      = One    -- represents 1
+      | Omega  -- represents ω
+
+- Accompanied by two specially recognised type families:
+
+  ::
+
+    type family (:+) :: Multiplicity -> Multiplicity -> Multiplicity
+    type family (:*) :: Multiplicity -> Multiplicity -> Multiplicity
+
+  Note: unification of
+  multiplicities will be performed up to the semiring laws for
+  ``(:+)`` and ``(:*)`` (see Specification_).
+
+The linear and unrestricted arrows are aliases:
+
+- ``(->)`` is an alias for ``(:'Omega ->)``
+- ``(->.)`` (ASCII syntax) and ``(⊸)`` (Unicode syntax) are aliases
+  for ``(:'One ->)``
+
+Constructors & pattern-matching
+-------------------------------
+
+Constructors of data types defined with the Haskell'98 syntax
+
+::
+
+  data Foo
+    = Bar A B
+    | Baz C
+
+have linear function types, that is ``Bar :: A ->. B ->. Foo``. This
+is true in every module, including those without ``-XLinearTypes``
+turned on. This implies that most types in ``base`` (``Maybe``,
+``[]``, etc…) have linear constructors. We also make the constructor
+of primitive tuples ``(,)`` linear in their arguments.
+
+With the GADT syntax, multiplicity of the arrows is honoured:
+
+::
+
+  data Foo2 where
+    Bar2 :: A ->. B -> Foo2
+
+means that ``Bar2 :: A ->. B -> Foo2``. This means that *data types
+written in GADT syntax with the ``(->)`` arrow are not the same as if
+they were defined with Haskell'98 syntax*.
```

Comment:
It would be good to clarify the behavior of GADT syntax when `-XLinearTypes` is disabled.

Reply:
Thanks for pointing it out. There is already a section dedicated to the [behaviour without `-XLinearTypes`](https://github.com/tweag/ghc-proposals/blob/linear-types2/proposals/0000-linear-types.rst#without-xlineartypes). It documents the behaviour of GADTs. However, I'll add a forward reference in this section.

- [ ] forward reference from GADT specification to behaviour without `-XLinearTypes`.